Framework/Algorithms/src/FitPeaks.cpp

```diff
@@ -870,74 +930,133 @@ std::vector<std::shared_ptr<FitPeaksAlgorithm::PeakFitResult>> FitPeaks::fitPeak
+/** Sum up all counts in a histogram
+ * @param histogram :: histogram instance
+ * @return :: total number of counts
+ */
 double numberCounts(const Histogram &histogram) {
-  double total = 0.;
-  for (const auto &value : histogram.y())
-    total += std::fabs(value);
+  double total = std::accumulate(histogram.y().begin(), histogram.y().end(), 0.);
   return total;
 }
 
 //----------------------------------------------------------------------------------------------
-/** Get number of counts in a specified range of a histogram
+/** Sum up all counts in a specified range of a histogram
  * @param histogram :: histogram instance
- * @param xmin :: left boundary
- * @param xmax :: right boundary
- * @return :: counts
+ * @param range :: histogram range
+ * @return :: total number of counts
  */
-double numberCounts(const Histogram &histogram, const double xmin, const double xmax) {
-  const auto &vector_x = histogram.points();
-
-  // determine left boundary
-  auto start_iter = vector_x.begin();
-  if (xmin > vector_x.front())
-    start_iter = std::lower_bound(vector_x.begin(), vector_x.end(), xmin);
-  if (start_iter == vector_x.end())
-    return 0.; // past the end of the data means nothing to integrate
-  // determine right boundary
-  auto stop_iter = vector_x.end();
-  if (xmax < vector_x.back()) // will set at end of vector if too large
-    stop_iter = std::lower_bound(start_iter, stop_iter, xmax);
-
-  // convert to indexes to sum over y
-  size_t start_index = static_cast<size_t>(start_iter - vector_x.begin());
-  size_t stop_index = static_cast<size_t>(stop_iter - vector_x.begin());
-
-  // integrate
-  double total = 0.;
-  for (size_t i = start_index; i < stop_index; ++i)
-    total += std::fabs(histogram.y()[i]);
+double numberCounts(const Histogram &histogram, const std::pair<double, double> &range) {
+  // get data range
+  std::vector<double> vec_x, vec_y, vec_e;
+  getRangeData(histogram, range, vec_x, vec_y, vec_e);
+  // sum up all counts
+  double total = std::accumulate(vec_y.begin(), vec_y.end(), 0.);
   return total;
 }
 
 //----------------------------------------------------------------------------------------------
-/** Get number of vector elements falling into specified range
- * @param vector :: vector of type T
+/** Estimate background noise given a peak-window Y-vector
+ * @param vec_y :: a peak window of Y-values
+ * @return :: noise estimate
+ */
+double estimateBackgroundNoise(const std::vector<double> &vec_y) {
+  // assume that the first and the last 10 data points are background. Combine them into a data vector.
+  assert(vec_y.size() > 20);
+  std::vector<double> vec_bkg;
+  vec_bkg.resize(20);
+  std::copy(vec_y.begin(), vec_y.begin() + 10, vec_bkg.begin());
+  std::copy(vec_y.end() - 10, vec_y.end(), back_inserter(vec_bkg));
+  // estimate the noise as the standard deviation of that combined data vector
+  auto intensityStatistics = Kernel::getStatistics(vec_bkg, StatOptions::CorrectedStdDev);
```

Comment:
I'm not sure whether this standard deviation approach is valid for Poisson distributed/low counts data (regime we are often in when calling `FitPeaks` from `PDCalibration`) - perhaps it would be better to use e.g. $(1/N_{bg})\sqrt{\Sigma_{i=0}^{N_{bg}} e_i^2}$ ? Just a suggestion, it may not make much difference!

Reply:
I believe the formula you gave is for a standard error of the mean (SEM), i.e the mean background value.  I am using a standard deviation of the population (SD), which is the background noise. SEM is normally less than SD.

Reply:
Yes I suggested the SEM because i thought it might be more robust for fewer bins but agreed it is not the 'noise' on the background - which is probably the important thing for excluding peaks.

Reply:
A criterion which a lot of crystallographers use to discriminate valid/real peaks from spurious peaks is the `I/sigma` ratio (i.e. the integrated intensity/counts divided by the error in the integrated intensity). This would require attempting to fit the peak (for which you need to be reasonably certain there may be a peak there - e.g. from looking at signal to noise ratio pre-fitting as done here), so maybe it is a separate check post-fitting? It's just an idea and would be done separately to this PR if at all.

Reply:
I agree, such a post-fit check would be quite useful. And yes, if approved, it would be done separately from this PR.